build_tools/index_generation_s3_tar.py

```diff
@@ -0,0 +1,181 @@
+#!/usr/bin/env python
+"""
+Script to generate an index.html listing .tar.gz files in an S3 bucket, performing the following:
+ * Lists .tar.gz files in the specified S3 bucket.
+ * Generates HTML page with sorting and filtering options
+ * Saves the HTML locally as index.html
+ * Uploads index.html back to the same S3 bucket
+
+Requirements:
+ * `boto3` Python package must be installed, e.g.: pip install boto3
+
+Examples:
+Generate index.html for all tarballs in a bucket:
+./index_generation_s3_tar.py --bucket <bucketname>
+
+
+Generate index.html for files under a prefix and in a specific region:
+./index_generation_s3_tar.py --bucket <bucketname> --prefix nightly/ --region us-east-1
+
+"""
+
+import os
+import argparse
+import boto3
+from botocore.exceptions import NoCredentialsError, ClientError
+import re
+import json
+from github_actions.github_actions_utils import gha_append_step_summary
+
+
+def extract_gpu_details(files):
+    gpu_family_pattern = re.compile(r"gfx(?:\d+[A-Za-z]*|\w+)")
+    gpu_families = set()
+    for file_name, _ in files:
+        match = gpu_family_pattern.search(file_name)
+        if match:
+            gpu_families.add(match.group(0))
+    return sorted(list(gpu_families))
+
+
+def generate_index_s3(s3_client, bucket_name):
+    # List all objects and select .tar.gz keys
+    try:
+        paginator = s3_client.get_paginator("list_objects_v2")
+        page_iterator = paginator.paginate(Bucket=bucket_name)
+    except NoCredentialsError:
+        raise Exception(
+            "AWS credentials not found"
+        )
+    except ClientError as e:
+        raise Exception(f"Error accessing bucket {bucket_name}: {e}")
+
+    files = []
+    for page in page_iterator:
+        for obj in page.get("Contents", []):
+            key = obj["Key"]
+            if key.endswith(".tar.gz"):
+                files.append((key, obj["LastModified"].timestamp()))
+
+    if not files:
+        raise Exception(f"No .tar.gz files found in bucket {bucket_name}.")
```

Comment:
Prefer specific Exception subclasses
```suggestion
    if not files:
        raise FileNotFoundError(f"No .tar.gz files found in bucket {bucket_name}.")
```